pkg/database/external_issuer_stats.go

```diff
@@ -0,0 +1,155 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package database
+
+import (
+	"bytes"
+	"encoding/csv"
+	"encoding/json"
+	"fmt"
+	"sort"
+	"strconv"
+	"time"
+
+	"github.com/google/exposure-notifications-verification-server/internal/icsv"
+)
+
+var _ icsv.Marshaler = (ExternalIssuerStats)(nil)
+
+// ExternalIssuerStats is a collection of external issuer stats.
+type ExternalIssuerStats []*ExternalIssuerStat
+
+// ExternalIssuerStat represents statistics related to a user in the database.
+type ExternalIssuerStat struct {
+	Date        time.Time `gorm:"column:date; type:date;"`
+	RealmID     uint      `gorm:"column:realm_id; type:int"`
+	IssuerID    string    `gorm:"column:issuer_id; type:varchar(255)"`
+	CodesIssued uint      `gorm:"column:codes_issued; type:int;"`
+}
+
+// MarshalCSV returns bytes in CSV format.
+func (s ExternalIssuerStats) MarshalCSV() ([]byte, error) {
+	// Do nothing if there's no records
+	if len(s) == 0 {
+		return nil, nil
+	}
+
+	var b bytes.Buffer
+	w := csv.NewWriter(&b)
+
+	if err := w.Write([]string{"date", "realm_id", "issuer_id", "codes_issued"}); err != nil {
+		return nil, fmt.Errorf("failed to write CSV header: %w", err)
+	}
+
+	for i, stat := range s {
+		if err := w.Write([]string{
+			stat.Date.Format("2006-01-02"),
+			strconv.FormatUint(uint64(stat.RealmID), 10),
+			stat.IssuerID,
+			strconv.FormatUint(uint64(stat.CodesIssued), 10),
+		}); err != nil {
+			return nil, fmt.Errorf("failed to write CSV entry %d: %w", i, err)
+		}
+	}
+
+	w.Flush()
+	if err := w.Error(); err != nil {
+		return nil, fmt.Errorf("failed to create CSV: %w", err)
+	}
+
+	return b.Bytes(), nil
+}
+
+type jsonExternalIssuerStat struct {
+	RealmID uint                           `json:"realm_id"`
+	Stats   []*jsonExternalIssuerStatStats `json:"statistics"`
+}
+
+type jsonExternalIssuerStatStats struct {
+	Date time.Time                     `json:"date"`
```

Comment:
maybe I'm tired or its my font, but "date" and "data" scan really similarly. I wonder if `issuerData` would be nice.

Reply:
Fixed